### PR TITLE
Inductor support for aten::all_reduce

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94440
* __->__ #93111

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -4015,3 +4015,106 @@ def debug_str(self, name="block"):
...
+        wrapper.writeline(f"{work}.wait()")
+
+        # largely symbolic from the runtime perspective, but we must ensure from scheduler
+        # perspective that Wait's buffer is the one being consumed by consumers, to ensure
+        # wait op gets scheduled at all.  And as a side effect, wait's buffer name needs to actually
+        # point to something in the codegen output
+
+        # this 'del' is screwing up scheduler 'reuse' of this buf. which i'm not sure should even be allowed
+        # wrapper.writeline(f"{self.get_name()} = {all_reduce}; del {all_reduce}")
+        wrapper.writeline(f"{self.get_name()} = {all_reduce}")
```
Comment (jansel) [inline torch/_inductor/ir.py:4050]: You might need to add this to `get_alias_names()` to prevent the other value from being reused.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -4015,3 +4015,106 @@ def debug_str(self, name="block"):
...
+        # Return a 'Wait' to the user that called 'all_reduce' in the first place.  It consumes the 'work'
+        # and waits on it, also producing a buffer which is really the input buffer to AllReduce.
+        return Wait(
+            layout=new_layout,
+            inputs=[all_reduce],
+        )
+
+    def codegen(self, wrapper):
+        # these header lines just get output once
+        wrapper.header.writeline("import torch.distributed as dist")
```
Comment (jansel) [inline torch/_inductor/ir.py:4097]: https://github.com/pytorch/pytorch/blob/661800a2cf0a4251c48a50fce8599c4c5fc49be1/torch/_inductor/codegen/wrapper.py#L319-L323

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -4015,3 +4015,106 @@ def debug_str(self, name="block"):
...
+        group_id = f"{repr(self.constant_args[0])}"
+        reduce_op = self.constant_args[1]
+        # TODO make this real
+        c10d_op = {"sum": "ReduceOp.SUM"}
+
+        # We must copy our input buffer sometimes, but the scheduler will help us find opportunities
+        # to reuse the input buffer.  (This requires no other users of the input buffer.)
+        sched_node = V.graph.scheduler.name_to_node[output_name]
+        input_read = next(iter(sched_node.read_writes.reads))
+        if not sched_node.can_inplace(input_read):
```
Comment (jansel) [inline torch/_inductor/ir.py:4112]: Double check to make sure this is robust.  Not totally convinced about correctness.

Comment (jansel) [reply torch/_inductor/ir.py:4112]: Might be better to move this flag to into the caller that might have more information.

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -224,6 +224,65 @@ def debug_str_extra(self):
     def is_extern(self):
         return True
 
+    def allocate(self):
```
Comment (jansel) [inline torch/_inductor/scheduler.py:227]: Deduplicate these, should be possible to share code

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py (hunk 3 above)
Comment (wconstab) [reply torch/_inductor/ir.py:4112]: by caller, you mean lowering.py? and which information would I be looking for?

I think my biggest issue with this is that "can_inplace" does not imply "did_inplace".  So I considered actually setting up a new map of 'things_inplaced" inside the scheduler that I could peek at here.  And perhaps if I set up the mutation thing correctly, I could just reuse the mutation map for this purpose (although I am not sure if that would lead to a false positive in this logic in some cases).

File: torch/_inductor/scheduler.py (hunk 4 above)
Comment (wconstab) [reply torch/_inductor/scheduler.py:227]: yep. First, I needed to resolve the difference- I removed some lines starting here from my copy
https://github.com/pytorch/pytorch/blob/master/torch/_inductor/scheduler.py#L331

namely, V.kernel in my case was a NullKernel or something, and had no attributes or functions such as 'make_inplace'. I don't know what kernel means in this case, especially if accessed through the V object. (Unclear how it relates to my current node).  I'll see if I can figure this out, and then if using the 'make_inplace' logic would actually help make my changes more robust.  Not sure if this is relevant at all.

File: torch/_inductor/ir.py (hunk 3 above)
Comment (jansel) [reply torch/_inductor/ir.py:4112]: Oh I was thinking of trying to get this in wrapper.py (where allocations happen)

File: torch/_inductor/ir.py (hunk 1 above)
Comment (kumpera) [reply torch/_inductor/ir.py:4050]: Could we just capture wait as part of tracing and make it functional, IE "wait(tensor) -> tensor".

Even better if we can capture it through AsyncCommTensor so it already lands on the right spot.

This way any code motion/scheduling pass would just work OOTB as we'd have captured the dependencies.

Comment (wconstab) [reply torch/_inductor/ir.py:4050]: well, we don't gain or lose information by having wait in our initial traced IR.

the question is, where do we want to do the scheduling?  so far, the plan from the discussions with folks is to try to do the scheduling in inductor.  sync with @lessw2020 also on this.

Comment (wconstab) [reply torch/_inductor/ir.py]: ok- we've changed to do things the way proposed by @kumpera, `Wait` is no longer added automatically by Inductor and is lowered separately from the input FX graph.

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -4202,3 +4200,105 @@ def debug_str(self, name="block"):
...
+    def codegen(self, wrapper):
+        (input_collective,) = [t.codegen_reference() for t in self.inputs]
+        work = f"{input_collective}_work"  # hacky way to name work objs..
+        wrapper.writeline(f"{work}.wait()")
+
+        # wait op still needs to produce a 'buffer' that represents the tensor output.
+        # this is a symbolic gesture, and it gets handled by WrapperCodegen.
+        # codegen outputs a '# reuse' line that assigns the input buffer here ('input_collective')
+        # to a new name (`self.get_name()`) and `del`s the old name.
+        wrapper.writeline(f"{self.get_name()} = {input_collective}")
```
Comment (jansel) [inline torch/_inductor/ir.py:4232]: Does the `input_collective` buffer name need to go in `get_alias_names()`?  I worry that after the wait inductor will think `input_collective` is free and reuse it for something else.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/ir.py (hunk 5 above)
Comment (wconstab) [reply torch/_inductor/ir.py:4232]: yea, i was able to confirm the bug with tests, and adding the input collective's name to `get_alias_names` fixed it.

Thanks!

File: test/distributed/test_traceable_collectives.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            # inside py registered collective ops
+            with enable_python_dispatcher():
+                eager_out = matmul_cat_col(*inputs)
+                compiled_matmul_cat_col = compile(matmul_cat_col, inputs)
+                inductor_out = compiled_matmul_cat_col(*inputs)
+                assert same(eager_out, inductor_out, tol=0.001)
+
+
+@requires_nccl()
+class TestCollectivesInductor(DynamoDistributedSingleProcTestCase):
```
Comment (wanchaol) [inline test/distributed/test_traceable_collectives.py:82]: pretty interested on how `SingleProcTestCase` works for collective, is it doing allreduce on a single rank?

File: test/distributed/test_traceable_collectives.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            ar = torch.ops.aten.all_reduce(inp, "sum", tag, ranks, group_size)
+            return ar
+
+        input = torch.ones(4, 4, device="cuda", requires_grad=True)
+        with enable_python_dispatcher():
+            # TODO implement backwards
+            with self.assertRaisesRegex(RuntimeError, "derivative for aten::all_reduce is not implemented"):
+                compiled = torch.compile(func, backend="aot_eager")  # inductor bug with single-op allreduce graph
+                out = compiled(input, **self.get_world_trs())
+                out.sum().backward()
```
Comment (wanchaol) [inline test/distributed/test_traceable_collectives.py:219]: Oh I thought we didn't implement the allreduce backward yet, so it's a dummy function right now and we just test the correctness of dummy function to see if it could work here?

File: torch/_inductor/ir.py [hunk 8]
```diff
@@ -4202,3 +4200,109 @@ def debug_str(self, name="block"):
...
+        # about that, and we just pretend for scheduling purposes that the work obj is a 1-elem tensor.
+        # Nobody should consume the output of AllReduce except 'Wait', which we control here.
+        return AllReduce(
+            layout=new_layout,
+            inputs=[x],
+            constant_args=[reduce_op, tag, ranks, group_size],
+        )
+
+    def codegen(self, wrapper):
+        wrapper.add_import_once("import torch.distributed as dist")
```
Comment (wanchaol) [inline torch/_inductor/ir.py:4279]: So right now it generates the triton python code I suppose, would it possible to generate a C++ kernel in the future?

File: test/distributed/test_traceable_collectives.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                .check("buf3 = buf1") \
+                .check("return (buf3, buf2, buf4") \
+                .run(code)
+            out = compiled(inputs, **self.get_world_trs())
+            correct = func(inputs, **self.get_world_trs())
+            assert same(out, correct)
+
+    def test_dynamo_trace_allreduce(self):
+        def func(inp, *, tag, ranks, group_size):
+            ar = torch.ops.aten.all_reduce(inp, "sum", tag, ranks, group_size)
```
Comment (wanchaol) [inline test/distributed/test_traceable_collectives.py:190]: dynamo works because we are calling the aten op not the functional collective directly, so we get around the AsyncTensor subclass?

**Review by wanchaol — APPROVED** [review]
lgtm, just some questions as I'm learning how inductor works and this looks a great example!

File: test/distributed/test_traceable_collectives.py (hunk 6 above)
Comment (wconstab) [reply test/distributed/test_traceable_collectives.py:82]: it is calling the allreduce op but maybe the nccl kernel is skipped.

these tests are only measuring whether inductor generates the right code to call into dist.* apis.

I assume the apis will work as intended.

Above there is one 'real' integration test that runs multi-proc

File: test/distributed/test_traceable_collectives.py (hunk 7 above)
Comment (wconstab) [reply test/distributed/test_traceable_collectives.py:219]: yea, i should delete this test.  I asked Rodrigo to cover this in his own test file, and he did.  Also, he dropped backward support for now so i think his test is a stub.  we will add it later.

File: torch/_inductor/ir.py [hunk 10]
```diff
@@ -4202,3 +4200,109 @@ def debug_str(self, name="block"):
...
+        wrapper.add_import_once(
+            "from torch.distributed._functional_collectives import _str_to_reduce_op"
+        )
+        wrapper.add_import_once(
+            "from torch.distributed.distributed_c10d import _find_or_create_pg_by_ranks_and_tag"
```
Comment (kumpera) [inline torch/_inductor/ir.py:4284]: I find using c10d internals a bit problematic but we can iterate over this later.

File: torch/_inductor/ir.py (hunk 8 above)
Comment (wconstab) [reply torch/_inductor/ir.py:4279]: all of the python code i'm generating here is _not_ triton code.  Triton is generated by one layer deeper of inductor, when it does a 'fusion' of some ops and then codegens a kernel.  This code here is going into the 'top level wrapper' script inductor generates, which is what calls the generated triton kernels and also calls other eager ops or allocations etc.

the python wrapper code can also be changed to c++, and that's part of the 'aot inductor' workstream.

File: torch/_inductor/ir.py [hunk 11]
```diff
@@ -4202,3 +4200,109 @@ def debug_str(self, name="block"):
...
+        )
+
+        # extract references to our args in string form for codegen output
+        (input_name,) = [t.codegen_reference() for t in self.inputs]
+        output_name = self.get_name()
+        reduce_op, tag, ranks, group_size = self.constant_args
+
+        # TODO: avoid more than one ref of the same pg (even though they are cached inside the api)
+        wrapper.writeline(
+            f"{output_name}_pg = _find_or_create_pg_by_ranks_and_tag('{tag}', {ranks}, {group_size})"
```
Comment (kumpera) [inline torch/_inductor/ir.py:4294]: This should be cached across invocations.

Comment (wconstab) [reply torch/_inductor/ir.py:4294]: yea, currently we will be constructing more than one obj that really are the same pg.  (and calling _find_or_create more than one time for the same pg)

is this a serious problem at all? I assumed it is 'safe' but also not ideal.  My todo above was framed as a cleanup for later.  But if you see a more serious issue let me know

File: torch/_inductor/scheduler.py [hunk 12]
```diff
@@ -313,6 +325,25 @@ def debug_str_extra(self):
     def is_extern(self):
         return True
 
+    def can_inplace(self, read_dep: dependencies.MemoryDep):
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_inductor/scheduler.py:328]: it's an 'interface' function already existing in inductor's scheduler.  i'm just defining specific behavior for this subclass

**Comment by wconstab** [comment]
pytorchbot merge -f "Flaky CI (unable to download huggingface model)"
